### PR TITLE
Azure Functions with Resteasy

### PR DESCRIPTION
**Pull request by patriot1burke**
1st iteration of Azure Functions with Resteasy.  Completely functional and documented!

### CODE REVIEW

**Review by gsmet — COMMENTED** [review]
I just did a quick check. Do we have someone familiar with Azure who could test that one?

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+If deployment is a success, the azure plugin will tell you the base URL to access your function.
+
+i.e.
+[source]
+----
+Successfully deployed the artifact to https://tutorials-resteasy-1.azurewebsites.net
+----
+
+The URL to access the service would be https://{azureFunctionAppName}.azurewebsites.net/api/hello
```
[the remark this answers is not in the document]
Comment (patriot1burke) [reply docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: It does not replace it.

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+
+Next create the azure maven project for your Quarkus application.
+
+[source, subs=attributes+]
+----
+mvn archetype:generate \
+    -DarchetypeGroupId=io.quarkus \
+    -DarchetypeArtifactId=quarkus-azure-functions-resteasy-archetype \
+    -DarchetypeVersion={quarkus-version}
```
[the remark this answers is not in the document]
Comment (patriot1burke) [reply docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: Take a look at the archetype.  The changes required in the pom are pretty complex.  A lot of copying and filtering.  A specific staging directory structure required by Azure.  Azure Descriptors that need to be editable and retained between build.  The need to add the azure-maven-plugin.

What I think I can do is reduce everything to one pom and just use the maven architetype to generate a resteasy example project.  I'll work on that now.

File: extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+import io.quarkus.deployment.annotations.Record;
+import io.quarkus.deployment.builditem.FeatureBuildItem;
+import io.quarkus.resteasy.server.common.deployment.ResteasyDeploymentBuildItem;
+
+public class AzureFunctionsResteasyProcessor {
+    private static final Logger log = Logger.getLogger(AzureFunctionsResteasyProcessor.class);
+
+    @BuildStep
+    FeatureBuildItem featureBuildItem() {
+        return new FeatureBuildItem("azure-resteasy");
```
[the remark this answers is not in the document]
Comment (patriot1burke) [reply extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java]: I don't understand this comment.

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (patriot1burke) [reply docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: What I mean is that the generated HTML looks fine.

File: extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gsmet) [reply extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java]: Sorry: for the other extensions, the name is declared as a constant in `FeatureBuildItem` - https://github.com/quarkusio/quarkus/blob/master/core/deployment/src/main/java/io/quarkus/deployment/builditem/FeatureBuildItem.java#L12 .

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (gsmet) [reply docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: Yes, I see why it is useful. I just remember a conversation where @cescoffier said archetypes wouldn't be flexible enough.

IIRC we have templates for projects in our tooling. Not sure if it is exposed. @cescoffier would know more about it.

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc (hunk 1 above)
[the remark this answers is not in the document]
Comment (gsmet) [reply docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: OK, thanks for checking.

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc (hunk 2 above)
[the remark this answers is not in the document]
Comment (patriot1burke) [reply docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: Yeah, I don't think I can get away from having two separate poms right now.  There's some issues with test/dev/deployment classloading.

File: extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (patriot1burke) [reply extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java]: Can we keep it out of there for now?  This extension will probably go away once we finish refactoring undertow and moving azure integration there.

[the remark this answers is not in the document]
Comment (gsmet) [reply extensions/azure-functions-resteasy/deployment/src/main/java/io/quarkus/azure/functions/resteasy/deployment/AzureFunctionsResteasyProcessor.java]: Yes, makes sense if we end up removing it.

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+== Prerequisites
+
+To complete this guide, you need:
+
+* less than 15 minutes
+* JDK 1.8 (Azure requires JDK 1.8)
+* Apache Maven 3.5.3+
+* A completed https://quarkus.io/getting-started[Getting Started Guide] for Quarkus.
+* An Azure Account
+* Azure CLI Installed
```
Comment (FroMage) [inline docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: A link to how to get the CLI would be helpful here.

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+
+== Prerequisites
+
+To complete this guide, you need:
+
+* less than 15 minutes
+* JDK 1.8 (Azure requires JDK 1.8)
+* Apache Maven 3.5.3+
+* A completed https://quarkus.io/getting-started[Getting Started Guide] for Quarkus.
+* An Azure Account
```
Comment (FroMage) [inline docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: Can you add a link to how we can get an account?

File: docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  this will be `getting-started`
+* version - Version that should match your existing quarkus application artifact.
+* package - use the default value
+* azureFunction - Name of your azure function. Can be anything you want.  Type in `getting-started`.
+* azureFunctionAppName - Application name of your azure function.  Must be unique subdomain name under *.azurewebsites.net. Otherwise deploying to Azure will fail.
+* azureFunctionAppRegion - Dependent on your azure region.  You can try `westus` if you don't know this value.
+* azureFunctionResourceGroup - Any value you want.
+
+This will generate a pom.xml that generates an artifactId with a suffix of `-azure`.  You can modify any of the values specified
+above in the pom.xml file, but you will need to completely rerun the build.
```
Comment (FroMage) [inline docs/src/main/asciidoc/azure-functions-resteasy-guide.adoc]: Shouldn't it be `pom.xml` (with backticks)? I think, the same could be said of every item name in the previous list, no?

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java [hunk 7]
```diff
@@ -240,24 +243,30 @@ public void build(
...
         Map<String, String> resteasyInitParameters = new HashMap<>();
+        ResteasyDeployment deployment = new ResteasyDeploymentImpl();
 
-        registerProviders(resteasyInitParameters, reflectiveClass, unremovableBeans, jaxrsProvidersToRegisterBuildItem);
+        registerProviders(deployment, resteasyInitParameters, reflectiveClass, unremovableBeans,
+                jaxrsProvidersToRegisterBuildItem);
 
         if (!resources.isEmpty()) {
+            deployment.getScannedResourceClasses().addAll(resources);
             resteasyInitParameters.put(ResteasyContextParameters.RESTEASY_SCANNED_RESOURCES, String.join(",", resources));
```
Comment (FroMage) [inline extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java]: do we really need to set both the init parameters and the deployment values?

**Comment by FroMage** [comment]
LGTM, but I did not follow the discussion around modules between @patriot1burke and @stuartwdouglas so I'll ask @stuartwdouglas to review.

File: extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java (hunk 7 above)
Comment (patriot1burke) [reply extensions/resteasy-server-common/deployment/src/main/java/io/quarkus/resteasy/server/common/deployment/ResteasyServerCommonProcessor.java]: Yes.  The init parameters will eventually disappear.  That's work that is scheduled.

**Comment by gsmet** [comment]
@patriot1burke Can you squash all the commits and force push?

Then we can wait for @stuartwdouglas 's opinion.

**Comment by patriot1burke** [comment]
> @patriot1burke Can you squash all the commits and force push?
> 
> Then we can wait for @stuartwdouglas 's opinion.

Ok, I screwed up the squash.  Squashed everything but top commit and thus the "files changed" are 1.3k files.  sigh.

**Comment by patriot1burke** [comment]
I'm just gonna create a new clean branch.

**Pull request closed without merging**